src/read.c

```diff
@@ -0,0 +1,214 @@
+#include "read.h"
+#include "framebuffer.h"
+#include "io_header.h"
+#include "macros.h"
+#include "parser.h"
+#include "utils.h"
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+void read_header(FILE *mib_ptr, unsigned long offset, framebuffer *fb)
+{
+  char *header;
+  MQ1_fields *mq1_header;
+  if (mib_ptr == NULL || fb == NULL) {
+    fprintf(stderr, "Missing input mib_ptr or fb in read_header\n");
+    return;
+  }
+  if (offset < 0) {
+    fprintf(stderr, "offset is negative, please check input\n");
+    return;
+  }
+  if (fseek(mib_ptr, offset, SEEK_SET) != 0) {
+    fprintf(stderr, "fseek error in read_header\n");
+    return;
+  }
+  char buf[MIB_HEADER_METADATA_BUF_SIZE]            = {0};
+  char headersize_str[MIB_HEADER_SIZE_FIELD_LENGTH] = {0};
+  size_t status =
+    fread(buf, sizeof(char), MIB_HEADER_METADATA_BUF_SIZE, mib_ptr);
+  if (status != MIB_HEADER_METADATA_BUF_SIZE) {
+    fprintf(stderr, "fread error in read_header\n");
+    return;
+  }
+  memcpy(headersize_str, buf + MIB_HEADER_SIZE_FIELD_OFFSET + 1,
+         MIB_HEADER_SIZE_FIELD_LENGTH - 1);
+  headersize_str[MIB_HEADER_SIZE_FIELD_LENGTH - 1] = '\0';
+  char *end_ptr;
+  long unsigned int headersize = strtoul(headersize_str, &end_ptr, 10);
+  if (end_ptr == headersize_str) {
+    fprintf(stderr, "headersize strtol error in read_header, no digit found\n");
+    return;
+  } else if (*end_ptr != '\0') {
+    fprintf(stderr,
+            "headersize strtol error in read_header, invalid character: %c\n",
+            *end_ptr);
+    return;
+  }
+  header = (char *) malloc(sizeof(char) * headersize);
```

Comment:
Being a paranoid person, I think we could include a check on the value of `headersize`. `headersize` is read from the file via `headersize_str` and potentially could be any value, if this happens to be a very large value for some reason, this will cause issues and hard to debug. We know currently the headersize could not be larger than `MQ1_QUAD_HEADER_BYTES` and we can utilise this information to perform a bound check.